File: docker_images/node/wrapper/glue/registryGlue.js
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
'use strict';
/*jshint esversion: 6 */

var debug = require('debug')('azure-iot-e2e:node')
var glueUtils = require('./glueUtils');
var NamedObjectCache = require('./NamedObjectCache');

var Registry = require('azure-iothub').Registry;


/**
 * cache of objects.  Used to return object by name to the caller.
 */
var objectCache = new NamedObjectCache();

/**
 * Connect to registry
 * Connect to the Azure IoTHub registry.  More specifically, the SDK saves the connection string that is passed in for future use.
 *
 * connectionString String connection string
 * returns connectResponse
 **/
exports.registry_Connect = function(connectionString) {
  debug(`registry_Connect called`);
  return glueUtils.makePromise('registry_Connect', function(callback) {
    var registry = Registry.fromConnectionString(connectionString);
    var connectionId = objectCache.addObject('registry', registry);
    callback(null, {connectionId: connectionId});
  });
}


/**
 * Disconnect from the registry
 * Disconnects from the Azure IoTHub registry.  More specifically, closes all connections and cleans up all resources for the active connection
 *
 * connectionId String Id for the connection
 * no response value expected for this operation
 **/
exports.registry_Disconnect = function(connectionId) {
  debug(`registry_Disconnect called with ${connectionId}`);
  return glueUtils.makePromise('registry_Disconnect', function(callback) {
    var registry = objectCache.removeObject(connectionId);
    if (!registry) {
      debug(`${connectionId} already closed.`);
      callback();
    } else {
      debug(`Removed registry for ${connectionId}.`);
      callback();
    }
  });
}


/**
 * gets the device twin for the given deviceid
 *
 * connectionId String Id for the connection
 * deviceId String 
 * returns Object
 **/
exports.registry_GetDeviceTwin = function(connectionId,deviceId) {
  return new Promise(function(resolve, reject) {
    glueUtils.returnFailure(reject);
  });
}


/**
 * gets the module twin for the given deviceid and moduleid
 *
 * connectionId String Id for the connection
 * deviceId String 
 * moduleId String 
 * returns Object
 **/
exports.registry_GetModuleTwin = function(connectionId,deviceId,moduleId) {
  debug(`registry_GetModuleTwin called with ${connectionId}, ${deviceId}, ${moduleId}`);
  return glueUtils.makePromise('registry_GetModuleTwin', function(callback) {
    var registry = objectCache.getObject(connectionId);
    debug(`calling Registry.getModuleTwin`);
    registry.getModuleTwin(deviceId, moduleId, function(err, result) {
      glueUtils.debugFunctionResult('registry.getModuleTwin', err);
      callback(err, result);
    });
  });
}


/**
 * update the device twin for the given deviceId
 *
 * connectionId String Id for the connection
 * deviceId String 
 * props Object 
 * no response value expected for this operation
 **/
exports.registry_PatchDeviceTwin = function(connectionId,deviceId,props) {
  return new Promise(function(resolve, reject) {
    glueUtils.returnFailure(reject);
  });
}


/**
 * update the module twin for the given deviceId and moduleId
 *
 * connectionId String Id for the connection
 * deviceId String 
 * moduleId String 
 * props Object 
 * no response value expected for this operation
 **/
exports.registry_PatchModuleTwin = function(connectionId,deviceId,moduleId,props) {
  debug(`registry_PatchModuleTwin called with ${connectionId}, ${deviceId}, ${moduleId}`);
  debug(props);
  return glueUtils.makePromise('registry_PatchModuleTwin', function(callback) {
    var registry = objectCache.getObject(connectionId);
    debug(`calling Registry.updateModuleTwin`);
    registry.updateModuleTwin(deviceId, moduleId, props, '*', function(err, result) {
      glueUtils.debugFunctionResult('registry.updateModuleTwin', err);
      callback(err, result);
    });
  });
}

File: docker_images/csharp/wrapper/src/Glue/RegistryGlue.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
using IO.Swagger.Models;
using Microsoft.Azure.Devices;
using Microsoft.Azure.Devices.Shared;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading.Tasks;

#pragma warning disable CA1304, CA1305, CA1307 // string function could vary with locale
#pragma warning disable CA1822 // method could be marked static
#pragma warning disable CA1822 // parameter is never used
#pragma warning disable CS1998 // async method lacks await operator

namespace IO.Swagger.Controllers
{
    /// <summary>
    /// Object which glues the swagger generated wrappers to the various IoTHub SDKs
    /// </summary>
    internal class RegistryGlue
    {
        private static Dictionary<string, object> objectMap = new Dictionary<string, object>();
        private static int objectCount = 0;
        private const string registryPrefix = "registry_";

        public RegistryGlue()
        {
        }

        public async Task<ConnectResponse> ConnectAsync(string connectionString)
        {
            Debug.WriteLine("RegistryConnectAsync called");
            var client = RegistryManager.CreateFromConnectionString(connectionString);
            Debug.WriteLine("Connecting registry manager object");
            await client.OpenAsync().ConfigureAwait(false);
            var connectionId = registryPrefix + Convert.ToString(++objectCount);
            Debug.WriteLine("Registry manager connection complete.  ConnectionId = " + connectionId);
            objectMap[connectionId] = client;
            return new ConnectResponse
            {
                ConnectionId = connectionId
            };
        }

        public async Task DisconnectAsync(string connectionId)
        {
            Debug.WriteLine("RegistryDisconnectAsync called for " + connectionId);
            if (objectMap.ContainsKey(connectionId))
            {
                var client = objectMap[connectionId] as RegistryManager;
                objectMap.Remove(connectionId);
                Debug.WriteLine("Calling CloseAsync on the RegistryManager object");
                await client.CloseAsync().ConfigureAwait(false);
                Debug.WriteLine("RegistryManager.CloseAsync complete");
            }
            else
            {
                Debug.WriteLine("registry object is already closed.");
            }
        }

        public async Task<Models.Twin> GetModuleTwin(string connectionId, string deviceId, string moduleId)
        {
            Debug.WriteLine("RegistryModuleTwinGet received for {0} with deviceId {1} and moduleId {2}", connectionId, deviceId, moduleId);
            var client = objectMap[connectionId] as RegistryManager;
            Debug.WriteLine("Getting twin");
            var twin = await client.GetTwinAsync(deviceId, moduleId).ConfigureAwait(false);
            Debug.WriteLine("Twin received.");
            Debug.WriteLine(JsonConvert.SerializeObject(twin));
            return new Models.Twin
            {
                Desired = twin.Properties.Desired,
                Reported = twin.Properties.Reported
            };
        }

        public async Task PatchModuleTwin(string connectionId, string deviceId, string moduleId, Models.Twin twin)
        {
            Debug.WriteLine("RegistryTwinPatchPutAsync received for {0} with deviceId {1} and moduleId {2}", connectionId, deviceId, moduleId);
            Debug.WriteLine(JsonConvert.SerializeObject(twin));
            var client = objectMap[connectionId] as RegistryManager;
            Debug.WriteLine("Patching twin");
            var registryTwin = new Microsoft.Azure.Devices.Shared.Twin {
                Properties = new TwinProperties {
                    Desired = new TwinCollection(twin.Desired as JObject, null)
                }
            };
            await client.UpdateTwinAsync(deviceId, moduleId, registryTwin, "*").ConfigureAwait(false);
            Debug.WriteLine("patch complete");
        }

        public async Task CleanupResourcesAsync()
        {
            if (objectMap.Count > 0)
            {
                string[] keys = new string[objectMap.Count];
                objectMap.Keys.CopyTo(keys, 0);
                foreach (var key in keys)
                {
                    await DisconnectAsync(key).ConfigureAwait(false);
                }
            }
        }


    }
}
